
<!DOCTYPE html
  PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head>
      <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
   <title>Flat Databases HOWTO</title><link rel="stylesheet" href="ede.css" type="text/css"><meta name="generator" content="DocBook XSL Stylesheets V1.61.3"><meta name="description" content="&#xA;	  The Open Biological Database Access (OBDA) standard&#xA;	  specifies a way of generating indexes for entry-based&#xA;	  sequence files (e.g. FASTA, EMBL) so that the entries can be&#xA;	  looked up and retrieved quickly.  These indexes are created&#xA;	  and accessed using the Bio::DB::Flat&#xA;	  module.&#xA;	"></head><body><div xmlns="" id="customheader">
BioPerl DocBook
</div><div class="article" lang="en"><div class="titlepage"><div><div><h1 class="title"><a id="d3e1"></a>Flat Databases HOWTO</h1></div><div><div class="author"><h3 class="author"><span class="firstname">Lincoln</span> <span class="surname">Stein</span></h3><div class="affiliation"><span class="orgname">
	  <a href="http://www.cshl.org" target="_top">Cold Spring Harbor Laboratory</a>
	<br></span><div class="address"><p><br>
	&nbsp;&nbsp;<tt class="email">&lt;<a href="mailto:lstein-at-cshl.org">lstein-at-cshl.org</a>&gt;</tt><br>
	</p></div></div></div></div><div><div class="author"><h3 class="author"><span class="firstname">Brian</span> <span class="surname">Osborne</span></h3><div class="affiliation"><span class="orgname"><a href="http://www.cognia.com" target="_top">Cognia Corporation</a><br></span><div class="address"><p><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<tt class="email">&lt;<a href="mailto:brian-at-cognia.com">brian-at-cognia.com</a>&gt;</tt><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</p></div></div></div></div><div><div class="author"><h3 class="author"><span class="firstname">Heikki</span> <span class="surname">Lehv&auml;slaiho</span></h3><div class="affiliation"><span class="orgname">
          <a href="http://www.ebi.ac.uk" target="_top">European Bioinformatics Institute</a>
        <br></span><div class="address"><p><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<tt class="email">&lt;<a href="mailto:heikki-at-ebi.co.uk">heikki-at-ebi.co.uk</a>&gt;</tt><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</p></div></div></div></div><div><div class="legalnotice"><p>This document is copyright Lincoln Stein, 2002.  For
           reproduction other than personal use please contact lstein at cshl.org
         </p></div></div><div><p class="pubdate">2003-02-26</p></div><div><div class="revhistory"><table border="1" width="100%" summary="Revision history"><tr><th align="left" valign="top" colspan="3"><b>Revision History</b></th></tr><tr><td align="left">Revision 1.0</td><td align="left">2003-02-26</td><td align="left">LS</td></tr><tr><td align="left" colspan="3">First version</td></tr><tr><td align="left">Revision 1.1</td><td align="left">2003-10-17</td><td align="left">BIO</td></tr><tr><td align="left" colspan="3"></td></tr><tr><td align="left">Revision 1.2</td><td align="left">2003-10-17</td><td align="left">HL</td></tr><tr><td align="left" colspan="3">from txt reformatted into SGML</td></tr><tr><td align="left">Revision 1.3</td><td align="left">2004-05-20</td><td align="left">BIO</td></tr><tr><td align="left" colspan="3">Add section on custom indexes</td></tr></table></div></div><div><div class="abstract"><p class="title"><b>Abstract</b></p><p>
	  The Open Biological Database Access (OBDA) standard
	  specifies a way of generating indexes for entry-based
	  sequence files (e.g. FASTA, EMBL) so that the entries can be
	  looked up and retrieved quickly.  These indexes are created
	  and accessed using the <tt class="classname">Bio::DB::Flat</tt>
	  module.
	</p></div></div></div><div></div><hr></div><div class="toc"><p><b>Table of Contents</b></p><dl><dt>1. <a href="#intro">Creating OBDA-Compliant Indexed Sequence Files</a></dt><dt>2. <a href="#options">Choosing Your Options</a></dt><dt>3. <a href="#files">Moving Database Files</a></dt><dt>4. <a href="#d3e160">Secondary or custom namespaces</a></dt><dt>5. <a href="#more">More information</a></dt></dl></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="intro"></a>1.&nbsp;Creating OBDA-Compliant Indexed Sequence Files</h2></div></div><div></div></div><p>
      <tt class="classname">Bio::DB::Flat</tt> has the same functionality as
      the various <tt class="classname">Bio::Index</tt> modules.  The main
      reason to use it is if you want to use the BioSequence Registry
      system (see the OBDA Access HOWTO at <a href="http://bioperl.org/HOWTOs" target="_top">http://bioperl.org/HOWTOs</a>),
      or if you want to share the same indexed files among scripts
      written in other languages, such as those written with BioJava or
      BioPython.
    </p><p>
      There are four steps to creating a
      <tt class="classname">Bio::DB::Flat</tt> database:
    </p><div class="orderedlist"><ol type="1"><li><p>
	  Select a Root Directory
	</p><p>
	  Select a directory in which the flat file indexes will be stored.
	  This directory should be writable by you, and readable by everyone
	  who will be running applications that access the sequence data.
	</p></li><li><p>
	  Move the Flat Files Into a Good Location
	</p><p>
	  The indexer records the path to the source files (e.g. FASTA, or
	  local copies of GenBank, Embl or SwissProt).  This means that you
	  must not change the location or name of the source files after
	  running the indexer.  Pick a good stable location for the source
	  files and move them there.
	</p></li><li><p>
	  Choose a Symbolic Name for the Database
	</p><p>
	  Choose a good symbolic name for the database.  For example, if you
	  are mirroring GenBank, "genbank" might be a good choice.  The
	  indexer will create files in a subdirectory by this name located
	  underneath the root directory.
	</p></li><li><p>
	  Run the bioflat_index.pl Script
	</p><p>
	  The final step is to run the bioflat_index.PLS script.  This
	  script is located in the BioPerl distribution, under scripts/DB.
	  For convenience, you are offered the option to copy it to
	  /usr/bin or another system-wide directory on 'make install' (and
	  its name will be changed to bioflat_index.pl).
	</p></li></ol></div></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="options"></a>2.&nbsp;Choosing Your Options</h2></div></div><div></div></div><p>
      The typical usage is as follows:
    </p><p>
      <pre class="programlisting">
	bioflat_index.pl -c -l /usr/share/biodb -d genbank -i bdb -f fasta data/*.fa
      </pre>
    </p><p>
     The following command line options are required:
      <div class="table"><a id="d3e86"></a><table summary="bioflat_index command-line options" border="1"><colgroup><col><col></colgroup><thead><tr><th>Option</th><th>Description</th></tr></thead><tbody><tr><td>-c</td><td>create a new index</td></tr><tr><td>-l</td><td>path to the root directory</td></tr><tr><td>-d</td><td>symbolic name for the new database</td></tr><tr><td>-i</td><td>indexing scheme (discussed below)</td></tr><tr><td>-f</td><td>source file format</td></tr></tbody></table><p class="title"><b>Table&nbsp;1.&nbsp;bioflat_index command-line options</b></p></div>
    </p><p>
      The <i class="parameter"><tt>-c</tt></i> option must be present to create the
      database.  If the database already exists,
      <i class="parameter"><tt>-c</tt></i> will reinitialize the index, wiping out
      its current contents.
    </p><p>
      The <i class="parameter"><tt>-l</tt></i> option specifies the root directory
      for the database indexes.
    </p><p>
      The <i class="parameter"><tt>-d</tt></i> option chooses the symbolic name for
      the new database.  If the <i class="parameter"><tt>-c</tt></i> option is
      specified, this will cause a new directory to be created
      underneath the root directory.
    </p><p>
      The <i class="parameter"><tt>-i</tt></i> option selects the indexing scheme.  
      Currently there are two indexing schemes supported: "bdb" and
      "flat."  "bdb" selects an index based on the Berkeley DB library.  
      It is generally the faster of the two, but it requires that the
      Berkeley DB library (from Linux RPM or from <a href="http://www.sleepycat.com" target="_top">www.sleepycat.com</a>, 
      version 2 or higher) and the Perl BerkeleyDB module be installed 
      on your system. The Perl DB_File module will not work.
    </p><p>
      "flat" is a sorted text-based index that uses a binary search algorithm to
      rapidly search for entries.  Although not as fast as bdb, the flat
      indexing system has good performance for even large databases, and
      it has no requirements beyond Perl itself. The disadvantage of
      "flat" is that the indexing of large databases will be slower and more
      memory-intensive than the indexing using "bdb".
    </p><p>
      Once an indexing
      scheme has been selected there is no way to change it other than
      recreating the index from scratch using the
      <i class="parameter"><tt>-c</tt></i> option.
    </p><p>
      The <i class="parameter"><tt>-f</tt></i> option specifies the format of the
      source database files.  It must be one of the many formats that BioPerl
      supports, including "genbank", "swiss", "embl" or "fasta".
      Consult the <tt class="classname">Bio::SeqIO</tt> documentation for
      the complete list.  All files placed in the index must share the
      same format.
    </p><p>
      The indexing script will print out a progress message every 1000
      entries, and will report the number of entries successfully
      indexed at the end of its run.
    </p><p>
      To update an existing index run bioflat_index.pl without the
      <i class="parameter"><tt>-c</tt></i> option and list the files to be added or
      reindexed.  The <i class="parameter"><tt>-l</tt></i> and
      <i class="parameter"><tt>-d</tt></i> options are required, but the indexing
      scheme and source file format do not have to be specified for
      updating as they will be read from the existing index.
    </p><p>
      For your convenience, bioflat_index.pl will also take default values
      from the following environment variables:
      <div class="table"><a id="d3e132"></a><table summary="Environmental variables" border="1"><colgroup><col><col></colgroup><thead><tr><th>ENV variable</th><th>description</th></tr></thead><tbody><tr><td>OBDA_FORMAT</td><td>format of sequence file (<i class="parameter"><tt>-f</tt></i>)
	      </td></tr><tr><td>OBDA_LOCATION</td><td>path to directory
		in which index files are stored
		(<i class="parameter"><tt>-l</tt></i>)
	      </td></tr><tr><td>OBDA_DBNAME</td><td>name of database (-d)</td></tr><tr><td>OBDA_INDEX</td><td>type of index to create (<i class="parameter"><tt>-i</tt></i>)</td></tr></tbody></table><p class="title"><b>Table&nbsp;2.&nbsp;Environmental variables</b></p></div>
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="files"></a>3.&nbsp;Moving Database Files</h2></div></div><div></div></div><p>
      If you must change the location of the source sequence files after
      you create the index, there is a way to do so.  Inside the root
      directory you will find a subdirectory named after the database,
      and inside that you will find a text file named "config.dat."  An
      example config.dat is shown here:
      <pre class="programlisting">
     index	flat/1
     fileid_0	/share/data/alnfile.fasta	294
     fileid_1	/share/data/genomic-seq.fasta	171524
     fileid_2	/share/data/hs_owlmonkey.fasta	416
     fileid_3	/share/data/test.fasta	804
     fileid_4	/share/data/testaln.fasta	4620
     primary_namespace	ACC
     secondary_namespaces	ID
     format	URN:LSID:open-bio.org:fasta
      </pre>
    </p><p>
      For each source file you have moved, find its corresponding
      "fileid" line and change the path.  Be careful not to change
      anything else in the file or to inadvertently replace tab
      characters with spaces.
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="d3e160"></a>4.&nbsp;Secondary or custom namespaces</h2></div></div><div></div></div><p>
      The bioflat_index script creates what could be called a default
      index using sequence files in a few different formats. Each of
      these formats has its own default primary key, the key that can
      be used as a a query. Consider this fasta entry:
    </p><p>
      <pre class="programlisting">
      &gt;P84139 gi|443893
      MHHLLHGRHHRRQRKKKITEWSVKKKIIACNMIIRRIIIIIAHYTRQWPLMNVDS
      </pre>
      </p><p>
      The Bio::DB::Flat modules will use the first "word" in the fasta
      header as the primary key, "P84139", equivalent to this regular expression:
    </p><p>
      <pre class="programlisting">
      &gt;(\S+)
      </pre>
    </p><p>
      This value turns out to be same value returned by Sequence
      object's display_id() method. The table below shows the 
      default primary keys of the 4 Bio::DB::Flat formats.
    </p><p>
      <div class="table"><a id="d3e170"></a><table summary="Default primary keys" border="1"><colgroup><col><col><col></colgroup><thead><tr><th>Format</th><th>Seq
		method</th><th>Regular expression</th></tr></thead><tbody><tr><td>fasta</td><td>display_id()</td><td>&gt;(\S+)</td></tr><tr><td>swiss</td><td>display_id()</td><td>^ID\s+(\S+)</td></tr><tr><td>genbank</td><td>primary_id()</td><td>^LOCUS\s+(\S+)</td></tr><tr><td>embl</td><td>display_id()</td><td>^ID\s+(\S+)</td></tr></tbody></table><p class="title"><b>Table&nbsp;3.&nbsp;Default primary keys</b></p></div>
    </p><p>
      What if you wanted to use some other part of the entry as a
      key, like the GI number? This could also be called specifying 
      another namespace, or a secondary namespace. No problem, but now 
      you've gone beyond the capabilities of the bioflat_index script, 
      you'll need to write your own. It would look something like this:
    </p><p>
      <pre class="programlisting">
      use strict;
      use Bio::DB::Flat::BinarySearch;

      # use single quotes so you don't have to write
      # regular expressions like "gi\\|(\\d+)"
      my $primary_pattern = '^&gt;(\S+)';
      # one or more patterns stored in a hash:
      my $secondary_patterns = {GI =&gt; 'gi\|(\d+)'};

      my $db = Bio::DB::Flat::BinarySearch-&gt;new(
                            -directory          =&gt; "/home/bio",
			    -dbname             =&gt; "ppp",
		            -write_flag         =&gt; 1,
			    -primary_pattern    =&gt; $primary_pattern,
			    -primary_namespace  =&gt; 'ACC',
			    -secondary_patterns =&gt; $secondary_patterns,
			    -verbose            =&gt; 1,
			    -format             =&gt; 'fasta'  );
      
      $db-&gt;build_index("ppp.fa");
      </pre>
    </p><p>
      This code will index the file "ppp.fa" and place the indexes in the 
      directory "/home/bio/ppp". Then you can retrieve sequences like this:
    </p><p>
      <pre class="programlisting">
      $db-&gt;secondary_namespaces("GI");
      my $acc_seq = $db-&gt;get_Seq_by_id("P84139");
      my $gi_seq = $db-&gt;get_Seq_by_secondary("GI",443893);
      </pre>
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="more"></a>5.&nbsp;More information</h2></div></div><div></div></div><p>
      For more information on using your indexed flat files please see the 
      <a href="http://bioperl.org/HOWTOs/OBDA_Access/index.html" target="_top">
	OBDA Access HOWTO</a>.
    </p></div></div><div xmlns="" id="customfooter">
BioPerl DocBook
</div></body></html>