
<!DOCTYPE html
  PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head>
      <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
   <title>Phylogenetic Tree HOWTO</title><link rel="stylesheet" href="ede.css" type="text/css"><meta name="generator" content="DocBook XSL Stylesheets V1.61.3"><meta name="description" content="&#xA;	This HOWTO intends to show how to use the Bioperl Tree objects to&#xA;	manipulate phylogenetic trees.  It shows how to read and write&#xA;	trees, query them for information about specific nodes or&#xA;	overall statistics.  Advanced topics include discussion of&#xA;	generating random trees and extensions of the basic structure&#xA;	for integration with other modules in Bioperl.&#xA;      "></head><body><div xmlns="" id="customheader">
BioPerl DocBook
</div><div class="article" lang="en"><div class="titlepage"><div><div><h1 class="title"><a id="d3e1"></a>Phylogenetic Tree HOWTO</h1></div><div><div class="author"><h3 class="author"><span class="firstname">Jason</span> <span class="surname">Stajich</span></h3><div class="affiliation"><span class="orgname">Dept Molecular Genetics and Microbiology,
	  Duke University<br></span><div class="address"><p><tt class="email">&lt;<a href="mailto:jason AT bioperl.org">jason&nbsp;AT&nbsp;bioperl.org</a>&gt;</tt></p></div></div></div></div><div><div class="legalnotice"><p>This document is copyright Jason Stajich, 2003.  It can
        be copied and distributed under the terms of the Perl
        Artistic License.
      </p></div></div><div><p class="pubdate">2003-12-01</p></div><div><div class="revhistory"><table border="1" width="100%" summary="Revision history"><tr><th align="left" valign="top" colspan="3"><b>Revision History</b></th></tr><tr><td align="left">Revision 0.1</td><td align="left">2003-12-01</td><td align="left">JES</td></tr><tr><td align="left" colspan="3">First version</td></tr></table></div></div><div><div class="abstract"><p class="title"><b>Abstract</b></p><p>
	This HOWTO intends to show how to use the Bioperl Tree objects to
	manipulate phylogenetic trees.  It shows how to read and write
	trees, query them for information about specific nodes or
	overall statistics.  Advanced topics include discussion of
	generating random trees and extensions of the basic structure
	for integration with other modules in Bioperl.
      </p></div></div></div><div></div><hr></div><div class="toc"><p><b>Table of Contents</b></p><dl><dt>1. <a href="#introduction">Introduction</a></dt><dt>2. <a href="#simple_usage">Simple Usage</a></dt><dt>3. <a href="#IO">Reading and Writing Trees</a></dt><dt>4. <a href="#d3e48">Example Code</a></dt><dt>5. <a href="#TreeI">Bio::Tree::TreeI methods</a></dt><dt>6. <a href="#TreeFunctionsI">Bio::Tree::TreeFunctionsI</a></dt><dt>7. <a href="#tree_building">Constructing Trees</a></dt><dt>8. <a href="#advanced_topics">Advanced Topics</a></dt><dt>9. <a href="#References">References and More Reading</a></dt></dl></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="introduction"></a>1.&nbsp;Introduction</h2></div></div><div></div></div><p>
      Generating and manipulating phylogenetic trees is an important
      part of modern systematics and molecular evolution research.
      The construction of trees is the subject of a rich literature
      and active research.  This HOWTO and the modules described
      within are focused on querying and manipulating trees once they
      have been created.      
    </p><p>
      The data we intend to capture with these objects concerns the notion
      of Trees and their Nodes.  A Tree is made up of Nodes and the
      relationships which connect these nodes.  The basic
      representation of parent and child nodes is intended to
      represent the directionality of evolution.  This is to capture
      the idea that some ancestral species gave rise, through speciation
      events, to a number of child species.  The data in the trees
      need not be a strictly bifurcating tree (or binary trees to the
      CS types), and a parent node can give rise to 1 or many child
      nodes.  
    </p><p>
      In practice there are just a few main
      objects, or modules, you need to know about.  There is the main
      Tree object <tt class="classname">Bio::Tree::Tree</tt> which is the main
      entry point to the data represented by a tree.  A Node is
      represented generically by <tt class="classname">Bio::Tree::Node</tt>,
      however there are subclasses of this object to handle particular
      cases where we need a richer object (see
      <tt class="classname">Bio::PopGen::Simulations::Coalescent</tt> and
      the PopGen HOWTO for more information).  The connections between
      Nodes are described using a few simple concepts. There is the concept of
      pointers or references where a particular Node keeps track of
      who its parent is and who its children are.  A Node can only have 1
      parent and it can have 1 or many children.  In fact all of the
      information in a tree pertaining to the relationships between
      Nodes and specific data, like bootstrap values and labels, are all
      stored in the Node objects while the
      <tt class="classname">Bio::Tree::Tree</tt> object is just a 
      container for some summary information about the tree and a
      description of the tree's root node.
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="simple_usage"></a>2.&nbsp;Simple Usage</h2></div></div><div></div></div><p>
      Trees are used to represent the history of a collection of taxa,
      sequences, or populations. 
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="IO"></a>3.&nbsp;Reading and Writing Trees</h2></div></div><div></div></div><p>
      Using <tt class="classname">Bio::TreeIO</tt> one can read trees
      from files or datastreams and
      create <tt class="classname">Bio::Tree::Tree</tt> objects.  This is
      analagous to how we read sequences from sequence files with
      <tt class="classname">Bio::SeqIO</tt> to create Bioperl sequence
      objects which can be queried and manipulated.  Similarly we can
      write <tt class="classname">Bio::Tree::Tree</tt> objects out to string
      representations like the Newick or New Hampshire format which can
      be printed to a file, a datastream, stored in database, etc.
    </p><p>
      The main module for reading and writing trees is the
      <tt class="classname">Bio::TreeIO</tt> factory module which has
      several driver modules which plug into it.  These drivers
      include <tt class="classname">Bio::TreeIO::newick</tt> for New
      Hampshire or Newick format,
      <tt class="classname">Bio::TreeIO::nhx</tt> for the New Hampshire eXtended
      format from Sean Eddy and Christian Zmeck as part of their RIO
      and ATV system [reference here].  The driver
      <tt class="classname">Bio::TreeIO::nexus</tt> supports parsing tree
      data from PAUP's Nexus format. However this driver currently
      only supports parsing, not writing, of Nexus format tree files.
   </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="d3e48"></a>4.&nbsp;Example Code</h2></div></div><div></div></div><p>
      Here is some code which will read in a Tree from a file called
      "tree.tre" and produce a Bio::Tree::Tree object which is stored in 
      the variable <pre class="programlisting">$tree</pre>.
    </p><p>
      Like most modules which do input/output you can also specify
      the argument -fh in place of -file to provide a glob or filehandle in
      place of the filename.  
    </p><p>
      <pre class="programlisting">
      use Bio::TreeIO;
      # parse in newick/new hampshire format
      my $input = new Bio::TreeIO(-file   =&gt; "tree.tre",
                                  -format =&gt; "newick");
      my $tree = $input-&gt;next_tree;			       
      </pre>
    </p><p>
      Once you have a Tree object you can do a number of things with it.
      These are all methods required in <tt class="classname">Bio::Tree::TreeI</tt>.
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="TreeI"></a>5.&nbsp;Bio::Tree::TreeI methods</h2></div></div><div></div></div><p>
      Request the taxa (leaves of the tree).
      <pre class="programlisting">my @taxa = $tree-&gt;get_leaf_nodes;</pre>
    </p><p>
      Get the root node.	
      <pre class="programlisting">my $root = $tree-&gt;get_root_node; </pre>
    </p><p>
      Get the total length of the tree (sum of all the branch lengths),
      which is only useful if the nodes actually have the branch
      length stored, of course.
      <pre class="programlisting">
my $total_length = $tree-&gt;total_branch_length;
      </pre>
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="TreeFunctionsI"></a>6.&nbsp;Bio::Tree::TreeFunctionsI</h2></div></div><div></div></div><p>
      An additional interface was written which implements 
      utility functions which are useful for manipulating a Tree.
    </p><p>
      Find a particular node, either by name or by some other field that is
      stored in a Node.  The field type should be the function name we
      can call on all of the Nodes in the Tree.
      <pre class="programlisting">
   # find all the nodes named 'node1' (there should be only one) 
   my @nodes = $tree-&gt;find_node(-id =&gt; 'node1');
   # find all the nodes which have description 'BMP'
   my @nodes = $tree-&gt;find_node(-description =&gt; 'BMP');
   # find all the nodes with bootstrap value of 70
   my @nodes = $tree-&gt;find_node(-bootstrap =&gt; 70);
      </pre>
    </p><p>
      If you would like to do more sophisticated searches, like "find all
     the nodes with bootstrap values better than 70", you can easily
     implement this yourself.
      <pre class="programlisting">
     my @nodes = grep { $_-&gt;bootstrap &gt; 70 } $tree-&gt;get_nodes;
      </pre>
      Remove a Node from the Tree and update the children/ancestor links
      where the Node is an intervening one.
    </p><p>
      <pre class="programlisting">
   # provide the node object to remove from the Tree
   $tree-&gt;remove_Node($node);
   # or specify the node Name to remove
   $tree-&gt;remove_Node('Node12');
      </pre>
    </p><p>
      Get the lowest common ancestor for a set of Nodes.  This method is
      used to find an internal Node of the Tree which can be traced,
      through its children, to the requested set of Nodes. It is used in
      the calculations of monophyly and paraphyly and in determining the
      distance between two nodes.
      <pre class="programlisting">
   # Provide a list of Nodes that are in this tree
   my $lca = $tree-&gt;get_lca(-nodes =&gt; \@nodes);
      </pre>
    </p><p>
      Get the distance between two nodes by adding up the branch lengths
      of all the connecting edges between two nodes.
      <pre class="programlisting">
   my $distances = $tree-&gt;distance(-nodes =&gt; [$node1,$node2]);
      </pre>
    </p><p>
   Perform a test of monophyly for a set of nodes and a given outgroup
   node.  This means the common ancestor for the members of the
   internal_nodes group is more recent than the common ancestor that any of them
   share with the outgroup node.
   <pre class="programlisting">
   if( $tree-&gt;is_monophyletic(-nodes    =&gt; \@internal_nodes,
                              -outgroup =&gt; $outgroup) ) {
     print "these nodes are monophyletic: ",
          join(",",map { $_-&gt;id } @internal_nodes ), "\n";
   }
      </pre>
    </p><p>
      Perform a test of paraphyly for a set of nodes and a given outgroup
      node.  This means that a common ancestor 'A' for the members of the
      ingroup is more recent than a common ancestor 'B' that they share with
      the outgroup node <span class="emphasis"><em>and</em></span> that there are no other 
      nodes in the tree which have 'A' as a common ancestor before 'B'.
      <pre class="programlisting">
   if( $tree-&gt;is_paraphyletic(-nodes    =&gt; \@internal_nodes,
                              -outgroup =&gt; $outgroup) &gt; 0 ) {
     print "these nodes are monophyletic: ",
          join(",",map { $_-&gt;id } @internal_nodes ), "\n";
   }   
      </pre>   
    </p><p>
   Reroot a tree, specifying a different node as the root (and a
   different node as the outgroup).
   <pre class="programlisting">
   # node can either be a Leaf node in which case it becomes the
   # outgroup and its ancestor is the new root of the tree
   # or it can be an internal node which will become the new
   # root of the Tree
   $tree-&gt;reroot($node);
      </pre>
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="tree_building"></a>7.&nbsp;Constructing Trees</h2></div></div><div></div></div><p>
      Pairwise distances for all sequences in an alignment can be
      computed with <tt class="classname">Bio::Align::DNAStatistics</tt> and
      (for DNA)
      and <tt class="classname">Bio::Align::ProteinStatistics</tt>.  There
      are several different methods implemented.  For DNA alignments,
      Jukes-Cantor (1969), Juke-Cantor uncorrected, Kimura 2-parameter
      (1980), Felsenstein (1981), Tajima-Nei (1984), and Tamura (1992)
      are currently implemented.  In addition, for coding sequences,
      synonymous and non-synonymous counts can be computed with
      the <tt class="function">calc_KaKs_pair</tt>.  For Protein sequences
      alignments only Kimura (1983) is currently supported although
      other methods will be added. 
    </p><p>
      To use these methods simply initialize a statistics module, and
      pass in an alignment object
      (<tt class="classname">Bio::SimpleAlign</tt>) and the type of distance
      method to use and the module will return
      a <tt class="classname">Bio::Matrix::PhylipDist</tt> matrix object of
      pairwise distances.  The code example below shows how this should
      be done.
    </p><p>
      Given the matrix of pairwise distances one can build a phylogenetic
      tree using 2 simple methods provided in
      the <tt class="classname">Bio::Tree::DistanceFactory</tt>.  Simple
      request either Neighbor-Joining (NJ) trees or Unweighted Pair Group
      Method with Arithmatic Mean (UPGMA) clusters.  There are caveats
      with these methods and whether or not the distances are additive.
      The method <tt class="function">check_additivity</tt> 
      in <tt class="classname">Bio::Tree::DistanceFactory</tt>
      is provided to calculate whether or not additivity holds for the
      data.
    </p><p> 
      The following is a basic code snippet which describes how to use
      the pairwise distance and tree building modules in Bioperl. 
    </p><p>
      <pre class="programlisting">
   use Bio::AlignIO;
   use Bio::Align::DNAStatistics;
   use Bio::Tree::DistanceFactory;
   # for a dna alignment
   # can also use ProteinStatistics
   my $aln = Bio::AlignIO-&gt;new(-file =&gt; 'filename', -format=&gt;'clustalw');
   my $stats = Bio::Align::DNAStatistics-&gt;new;
   my $mat = $stats-&gt;distance(-method =&gt; 'Kimura',
			      -align  =&gt; $aln);
   my $dfactory = Bio::Tree::DistanceFactory-&gt;new(-method =&gt; 'NJ');
   my $tree = $dfactory-&gt;make_tree($mat);
      </pre>
    </p><p>
      TODO: Using external programs: phylip,MrBayes,paup,puzzle,protml
    </p><p>
      Non-parametric bootstrapping is one method to test the consistency
      of the data with the optimal tree.  A set of subreplicates are
      generated from the alignment using the method
      from <tt class="classname">Bio::Align::Utilities</tt> called
      <tt class="function">bootstrap_replicates</tt>.  One passes in an
      alignment object and the count of the number of replicates to generate.
    </p><p>
      <pre class="programlisting">
   use Bio::Align::Utilities qw(:all); 
   my $replicates = bootstrap_replicates($aln,$count);
      </pre>
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="advanced_topics"></a>8.&nbsp;Advanced Topics</h2></div></div><div></div></div><p>
      It is possible to generate random tree topologies with a Bioperl
      object called <tt class="classname">Bio::Tree::RandomFactory</tt>.  The
      factory only requires the specification of the total number of taxa
      in order to simulate a history.  One can request different methods for
      generating the random phylogeny. At present, however, only the
      simple Yule backward is implemented and is the default. 
    </p><p>
      The trees can be generated with the following code.  You can either
      specify the names of taxa or just a count of total number of taxa
      in the simulation.
    </p><p>
      <pre class="programlisting">
   use Bio::TreeIO;
   use Bio::Tree::RandomFactory;
   # initialize a TreeIO writer to output the trees as we create them
   my $out = Bio::TreeIO-&gt;new(-format =&gt; 'newick',
                              -file   =&gt; "&gt;randomtrees.tre");
   my @listoftaxa = qw(A B C D E F G H);			      
   my $factory = new Bio::Tree::RandomFactory(-taxa =&gt; \@listoftaxa);
   # generate 10 random trees
   for( my $i = 0; $i &lt; 10; $i++ ) {
      $out-&gt;write_tree($factory-&gt;next_tree);
   }
   # One can also just request a total number of taxa (8 here) and 
   # not provide labels for them
   # In addition one can specify the total number of trees
   # the object should return so we can call this in a while
   # loop
   $factory = new Bio::Tree::RandomFactory(-num_taxa =&gt; 8
                                           -max_count=&gt; 10);
   while( my $tree = $factory-&gt;next_tree) {
     $out-&gt;write_tree($tree);
   }
      </pre>
    </p><p>
      There are more sophisticated operations that you may wish to pursue
      with these objects.  We have tried to create a framework for this type
      of data, but by no means should this be looked at as the final
      product.  If you have a particular statistic or function that
      applies to trees that you would like to see included in the
      toolkit we encourage you to send details to the Bioperl list.
    </p></div><div class="section" lang="en"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a id="References"></a>9.&nbsp;References and More Reading</h2></div></div><div></div></div><p>
      For more reading and some references for the techniques above see
      these titles.   
    </p><p>
      <table class="simplelist" border="0" summary="Simple list"><tr><td>
	  J. Felsenstein, "Infering Phylogenies" 2003. Sinuar and Associates.
	</td></tr><tr><td>
	  D. Swoffrod, Olsen, Waddell and D. Hillis, "Phylogenetic Inference"
	  1996. in Mol. Systematics, 2nd ed, 1996, Ch 11.
	</td></tr><tr><td>
	  Eddy SR, Durbin R, Krogh A, Mitchison G, "Biological Sequence
	  Analysis" 1998. Cambridge Univ Press, Cambridge, UK.
	</td></tr></table>
    </p></div></div><div xmlns="" id="customfooter">
BioPerl DocBook
</div></body></html>